File: cokoon/PyCokoonEditor/src/thememodel.py
'''
 Cokoon theme editor
 Copyright (C) 2007 Sandro Giessl <giessl@kde.org>

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; see the file COPYING.  If not, write to
 the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 Boston, MA 02110-1301, USA.
'''

from PyQt4 import QtCore
from PyQt4.QtXml import QDomDocument
from PyQt4.QtCore import QFile, QTextStream, QAbstractItemModel

from themespec import ThemeSpec, ThemeSpecDocument

class InvalidCokoonThemeException(Exception):
  '''Exception raised when an error occurs while handling a Cokoon Theme'''
  def __init__(self, msg, domElement):
    Exception.__init__(self)
    self.msg = msg
    self.domElement = domElement

class CokoonDomNode:
  def __init__(self, node, parent = None):
    '''initialises the node. with dom (QDomNode), 
    and a parent node'''
    self.node = node
    self.row = -1
    self.parent = parent
    self._childItems = None
  def setModified(self):
    self.parent.setModified()
  def isNull(self):
    return self.node.isNull()
  def children(self):
    '''returns the list of child nodes'''
    if self._childItems == None:
      self._updateChildItems()
    return self._childItems
  def child(self, i):
    '''returns the i-th child element'''
    if self._childItems == None:
      self._updateChildItems()
    if i >= 0 and len(self._childItems)>i:
      return self._childItems[i]
    return None
  def _updateChildItems(self):
    '''update the cache of child items'''
    self._childItems = []
    child = self.node.firstChild()
    count = 0
    while not child.isNull():
      n = self.createChildNode(child)
      if n:
        n.row = count
        self._childItems.append(n)
        ++count;
      child = child.nextSibling()

  def createChildNode(self, node):
    '''factory method which can be used by subclasses
    to implement the child() functionality.
    returns a CokoonDomNode if the node is suitable as a child, 
    or None if the node should be ignored'''
    if node.isComment():
      return CokoonDomComment(node, self)
    return None

  def id(self):
    return self.node.nodeName()
  def displayId(self):
    return self.id()
  def displayValue(self):
    attributes = QtCore.QStringList()
    attributeMap = self.node.attributes()
    for i in range(0, attributeMap.count()):
      attribute = attributeMap.item(i)
      attributes.append(attribute.nodeName() + "=\"" + \
                        attribute.nodeValue() + "\"")

    return attributes.join(" ")

class CokoonDomComment(CokoonDomNode):
  def __init__(self, node, parent):
    CokoonDomNode.__init__(self, node, parent)
    if not node.isComment():
      raise InvalidCokoonThemeException('QDomComment expected', self)
  def text(self):
    return self.node.nodeValue()
#    TODO: implement comment.
    return '...comment...';
  def displayValue(self):
    return self.text().split("\n").join(" ")

class CokoonDomElement(CokoonDomNode):
  def __init__(self, node, parent, tagName):
    CokoonDomNode.__init__(self, node, parent)
    if not node.isElement():
      raise InvalidCokoonThemeException('QDomDocument expected', self)
    self.element = node.toElement()
    self.tag = tagName
    if tagName and self.element.tagName() != tagName:
#      print 'tag %s expected,  but got %s'%(tagName, self.element.tagName())
      raise InvalidCokoonThemeException('tag %s expected'%(tagName, ), self)
  def _domAttr(self, attributeName):
    '''helper method which tries to extract the attribute attributeName
    and throws an InvalidCokoonThemeException if no such attribute exists'''
    attr = self.element.attribute(attributeName)
    if attr==None:
      raise InvalidCokoonThemeException('%s attribute %s not specified'%(self.tag, attributeName), self)
    return attr
  def _setDomAttr(self,attributeName,text):
    print "_setDomAttr",attributeName,":",text
    self.element.setAttribute(attributeName,text)
    self.setModified()

class CokoonThemeSource(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'source')
  def id(self):   return self._domAttr('id')
  def type(self): return self._domAttr('type')
  def file(self): return self._domAttr('file')
#  TODO: <replace_style_attr element_id="rect2788" attr="fill" value="ColorTitleBar"/>...

class CokoonThemeExpression(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'expression')
  def id(self):    return self._domAttr('id')
  def value(self): return self._domAttr('value')

class CokoonThemeTile(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'tile')
  def id(self):    return self._domAttr('id')
  def source_id(self): return self._domAttr('source_id')
  def mode_hor(self): return self._domAttr('mode_hor')
  def mode_vert(self): return self._domAttr('mode_vert')
  def left(self): return self._domAttr('left')
  def top(self): return self._domAttr('top')
  def width(self): return self._domAttr('width')
  def height(self): return self._domAttr('height')

class CokoonThemeLayer(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'layer')
  def layout_id(self):    return self._domAttr('layout_id')

class CokoonThemeLayout(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'layout')
  def id(self):    return self._domAttr('id')
#  TODO: column,  row,  cell

class CokoonThemePaint(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'paint')
  def createChildNode(self, node):
    '''reimplemented from CokoonDomNode'''
    if node.isElement():
      tag = node.toElement().tagName()
      if tag == 'layer':
        return CokoonThemeLayer(node, self)
    return CokoonDomElement.createChildNode(self, node)
#
#  def layers(self):
#    items = []
#    child = self.node.firstChildElement('layer')
#    while not child.isNull():
#      tag = child.tagName()
#      if tag=='layer':
#        items.append(CokoonThemeLayer(child, self))
#    return items

class CokoonThemeObject(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'object')
  def id(self):
    return self._domAttr('id')
  def createChildNode(self, node):
    '''reimplemented from CokoonDomNode'''
    if node.isElement():
      tag = node.toElement().tagName()
      if tag=='source':
        return CokoonThemeSource(node, self)
      elif tag=='expression':
        return CokoonThemeExpression(node, self)
      elif tag=='tile':
        return CokoonThemeTile(node, self)
      elif tag=='layout':
        return CokoonThemeLayout(node, self)
      elif tag=='paint':
        return CokoonThemePaint(node, self)
    return CokoonDomElement.createChildNode(self, node)

  def sources(self):
    srcs = []
    children = self.children()
    for child in children:
      if child.__class__ == CokoonThemeSource:
        srcs.append(child)
    return srcs

class CokoonTheme(CokoonDomElement):
  def __init__(self, node, parent):
    CokoonDomElement.__init__(self, node, parent, 'cokoon_theme')
  def name(self):
    return self._domAttr('name')
  def spec(self):
    return self._domAttr('spec')
  def version(self):
    return self._domAttr('version')
  def createChildNode(self, node):
    '''reimplemented from CokoonDomNode'''
    if node.isElement():
      tag = node.toElement().tagName()
      if tag=='object':
        return CokoonThemeObject(node, self)
    return CokoonDomElement.createChildNode(self, node)

  def objects(self):
    objs = []
    for child in self.children():
      if child.__class__ == CokoonThemeObject:
        objs.append(child)
    return objs

#  def object(self, objId):
#    for o in self.objects():
#      if o.id == objId:

class ThemeModel(QAbstractItemModel):
  '''This model implements a tree-like model which
  represents a Cokoon theme.'''

  def __init__(self):
    QAbstractItemModel.__init__(self, )
    self.doc = QDomDocument()
    self.currentFileName = None
    self.rootElement = None
    self._modified = False

    self.spec = None
    self.cokoonDoc = None

  def loadFile(self, file):
    '''Loads a theme description file (must be QFile)'''
    self.currentFileName = file.fileName()
    if not self.doc.setContent(file):
      self.clear()
      return False

    
    ## theme model initialization
    self.rootElement = CokoonTheme(self.doc.firstChildElement(), self)
    self.rootElement.row = 0

    ## make sure there is always a corresponding theme specification
    # TODO: do not hardcode the spec directory path...
    self.spec = ThemeSpec("/usr/share/apps/cokoon/specs/" + self.theme().spec() + ".xml")
    self.cokoonDoc = ThemeSpecDocument(self.spec)
    self.cokoonDoc.loadTheme(self.doc.toString(), file.fileName())

    ## theme modifications...
    self._modified = False
    
    return True

  def theme(self):
    return self.rootElement

  def clear(self):
    '''Clears the current document'''
    self.doc.setContent("")
    self.rootElement = None

  def save(self, file):
    '''Saves the theme description to file (must be QFile)'''
    outstr = QTextStream(file)
    outstr << self.doc.toString()

  def setModified(self, mod=True):
    self._modified = True

    if self.doc != None and self.currentFileName != None:
      self.cokoonDoc.clear()
      self.cokoonDoc.loadTheme(self.doc.toString(), self.currentFileName)

    self.emit(QtCore.SIGNAL("modelWasModified"), ())
    self.emit(QtCore.SIGNAL("dataChanged(const QModelIndex&,const QModelIndex&)"), QtCore.QModelIndex(),QtCore.QModelIndex())

  def isModified(self):
    '''Whether the document has been modified'''
    return self._modified

  def flags(self, index):
    '''reimplemented from QAbstractItemModel'''
    if not index.isValid():
      return QtCore.Qt.ItemIsEnabled
    
    return QtCore.Qt.ItemIsEnabled | QtCore.Qt.ItemIsSelectable

  def headerData(self, section, orientation, role):
    if orientation == QtCore.Qt.Horizontal and role == QtCore.Qt.DisplayRole:
      if section == 0:
          return QtCore.QVariant(self.tr("Item"))
      elif section == 1:
          return QtCore.QVariant(self.tr("ID"))
      elif section == 2:
          return QtCore.QVariant(self.tr("Value"))
      else:
          return QtCore.QVariant()
    
    return QtCore.QVariant()

  def data(self, index, role):
    '''reimplemented from QAbstractItemModel'''
    if not index.isValid():
      return QtCore.QVariant()
    if role != QtCore.Qt.DisplayRole:
      return QtCore.QVariant()
    
    item = index.internalPointer()

    node = item.node

    if index.column() == 0:
      return QtCore.QVariant(node.nodeName())
    
    elif index.column() == 1:
      return QtCore.QVariant(item.displayId() )
    elif index.column() == 2:
      return QtCore.QVariant(item.displayValue() )

    else:
      return QtCore.QVariant()

  def index(self, row, column, parent):
    '''reimplemented from QAbstractItemModel'''
    if not parent.isValid():
      parentItem = self.rootElement
    else:
      parentItem = parent.internalPointer()

    childItem = parentItem.child(row)
    if childItem:
      return self.createIndex(row, column, childItem)
    else:
      return QtCore.QModelIndex()

  def parent(self, child):
    '''reimplemented from QAbstractItemModel'''
    if not child.isValid():
      return QtCore.QModelIndex()
    
    childItem = child.internalPointer()
    parentItem = childItem.parent
    
    if not parentItem or parentItem == self.rootElement:
      return QtCore.QModelIndex()
    
    return self.createIndex(parentItem.row, 0, parentItem)

  def columnCount(self, parent):
    '''reimplemented from QAbstractItemModel'''
    return 3

  def rowCount(self, parent):
    '''reimplemented from QAbstractItemModel'''
    if not parent.isValid():
      parentItem = self.rootElement
    else:
      parentItem = parent.internalPointer()
    
    return len(parentItem.children())

File: cokoon/cokoon/themeeditor/thememodel.cpp
/* This file is part of the Cokoon theme library
 * Copyright (C) 2007 Sandro Giessl <giessl@kde.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <QDebug>

#include <QtAlgorithms>
#include <QHeaderView>

#include <QFile>
#include <QTextStream>
#include <kstandarddirs.h>

#include "../specification.h"
#include "thememodel.h"
#include "documentspecificationdocument.h"
#include <kmessagebox.h>
#include <kdirwatch.h>
#include <klocale.h>

namespace Cokoon {

    ThemeDomNode::ThemeDomNode(const QDomNode &node, const ThemeDomNode *parent)
        : m_node(node), m_parent(parent), m_theme(parent->themeModel())
    {
        qDebug() << "ThemeDomNode(..,..,parent:" << parent << "parent->thememodel:" << parent->themeModel();
    }

    ThemeDomNode::ThemeDomNode(const QDomNode &node, const ThemeDomNode *parent, ThemeModel *thememodel)
        : m_node(node), m_parent(parent), m_theme(thememodel)
    {
        qDebug() << "ThemeDomNode(..,..,themeModel:" << thememodel;
    }

    ThemeDomNode::~ThemeDomNode()
    {
        qDeleteAll(m_children);
    }

    QList<ThemeDomNode*> ThemeDomNode::children() const
    {
        if (m_children.isEmpty()) {
            QDomNode child = domNode().firstChild();
            int count = 0;
            while (!child.isNull()) {
                ThemeDomNode *n = createChildNode(child);
                if (n!=0) {
//                     qDebug() << " child" << n->displayText();
                    m_children.append(n);
                    ++count;
                }
                child = child.nextSibling();
            }
        }
        return m_children;
    }

    const ThemeDomNode *ThemeDomNode::parent() const
    {
        return m_parent;
    }

    ThemeModel *ThemeDomNode::themeModel() const
    {
        return m_theme;
    }

    QDomNode ThemeDomNode::domNode() const
    {
        return m_node;
    }

    QString ThemeDomNode::displayId() const
    {
        return "";
    }

    ThemeDomNode *ThemeDomNode::createChildNode(const QDomNode &node) const
    {
        if (node.isComment()) {
            return new ThemeComment(node, this);
        }
        return 0;
    }

    QDomNode ThemeDomNode::deleteChild(int index)
    {
        ThemeDomNode *n = children().value(index);
        if (n) {
            themeModel()->beginRemoveRows(modelIndex(),index,index);
            QDomNode old = domNode().removeChild(n->domNode());
            invalidateCache(); n=0; // also deletes n
            themeModel()->endRemoveRows();
            return old;
        }
        return QDomNode();
    }
    bool ThemeDomNode::insertChild(int index, const QDomNode &node)
    {
        if (!createChildNode(node)) {
            return false;
        }

        ThemeDomNode *n = children().value(index);
        bool ret;
        if (n) {
            themeModel()->beginInsertRows(modelIndex(),index,index);
            ret = !domNode().insertBefore(node, n->domNode()).isNull();
            invalidateCache();
            themeModel()->endInsertRows();
        } else {
            int mdlIdx = children().size()+1;
            themeModel()->beginInsertRows(modelIndex(),mdlIdx,mdlIdx);
            ret = !domNode().appendChild(node).isNull();
            invalidateCache();
            themeModel()->endInsertRows();
        }
        themeModel()->setModified(true);
        return ret;
    }

    void ThemeDomNode::invalidateCache()
    {
        qDeleteAll(m_children);
        m_children.clear();
    }

    QString ThemeDomNode::nodePath() const
    {
        ThemeDomNode *p = (ThemeDomNode*) parent();
        if (p) {
            int idx = p->children().indexOf((ThemeDomNode*)this);
            return QString("%1/%2").arg(p->nodePath()).arg(idx);
        } else {
            return "";
        }
    }

    int ThemeDomNode::childIndex() const
    {
        if (parent()) {
            return parent()->children().indexOf((ThemeDomNode*)this);
        } else {
            return 0;
        }
    }

    QModelIndex ThemeDomNode::modelIndex(int column) const
    {
        return themeModel()->createIndex(childIndex(), column, (void*)this);
    }

    ThemeDomElement::ThemeDomElement(const QDomNode &node, const ThemeDomNode *parent, const QString &tag)
        : ThemeDomNode(node, parent)
    {
        Q_ASSERT(node.isElement());
        Q_ASSERT(tag == domElement().tagName());
    }

    ThemeDomElement::ThemeDomElement(const QDomNode &node, const ThemeDomNode *parent, ThemeModel *thememodel, const QString &tag)
        : ThemeDomNode(node, parent, thememodel)
    {
        Q_ASSERT(node.isElement());
        Q_ASSERT(tag == domElement().tagName());
    }

    QString ThemeDomElement::displayText() const {
        return domElement().tagName();
    }
    QString ThemeDomElement::displayId() const {
        return getAttr("id");
    }
    QString ThemeDomElement::getAttr(const QString &attrName) const {
        return domElement().attribute(attrName);
    }
    void ThemeDomElement::removeAttr(const QString &attrName)  {
        domElement().removeAttribute(attrName);
        themeModel()->setThemeNodeEdited(nodePath());
    }
    void ThemeDomElement::setAttr(const QString &attrName, const QString &value) {
        domElement().setAttribute(attrName,value);
        themeModel()->setThemeNodeEdited(nodePath());
    }

    QDomElement ThemeDomElement::domElement() const
    {
        return domNode().toElement();
    }

    ThemeComment::ThemeComment(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomNode(node, parent) {
        Q_ASSERT(node.isComment());
    }
    ThemeSource::ThemeSource(const QDomNode &node, const ThemeDomNode *parent)
        : ThemeDomElement(node, parent, "source") {}
    ThemeSourceReplace::ThemeSourceReplace(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "replace_style_attr") {}
    ThemeExpression::ThemeExpression(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "expression") {}
   ThemeTile::ThemeTile(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "tile") {}
    ThemeLayer::ThemeLayer(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "layer") {}
    ThemeLayoutCell::ThemeLayoutCell(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "cell") {}
    ThemeLayoutColumn::ThemeLayoutColumn(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "column") {}
    ThemeLayoutRow::ThemeLayoutRow(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "row") {}
    ThemeLayout::ThemeLayout(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "layout") {}
    ThemePaint::ThemePaint(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "paint") {}
    ThemeObject::ThemeObject(const QDomNode &node, const ThemeDomNode *parent)
     : ThemeDomElement(node, parent, "object") {}

    Theme::Theme(const QDomNode &node, ThemeModel *model)
        : ThemeDomElement(node, 0, model, "cokoon_theme")
    {
    }


    ThemeDomNode *ThemeLayoutRow::createChildNode(const QDomNode &node) const {
        if (node.isElement() && node.toElement().tagName()=="cell")
                return new ThemeLayoutCell(node, this);
        return ThemeDomElement::createChildNode(node);
    }
    ThemeDomNode *ThemeLayout::createChildNode(const QDomNode &node) const {
        if (node.isElement()) {
            if (node.toElement().tagName()=="column")
                return new ThemeLayoutColumn(node, this);
            else if (node.toElement().tagName()=="row")
                return new ThemeLayoutRow(node, this);
        }
        return ThemeDomElement::createChildNode(node);
    }
    ThemeDomNode *ThemeObject::createChildNode(const QDomNode &node) const
    {
        if (node.isElement()) {
            if (node.toElement().tagName()=="source")
                return new ThemeSource(node, this);
            else if (node.toElement().tagName()=="expression")
                return new ThemeExpression(node, this);
            else if (node.toElement().tagName()=="tile")
                return new ThemeTile(node, this);
            else if (node.toElement().tagName()=="layout")
                return new ThemeLayout(node, this);
            else if (node.toElement().tagName()=="paint")
                return new ThemePaint(node, this);
        }
        return ThemeDomElement::createChildNode(node);
    }
    ThemeDomNode *ThemeSource::createChildNode(const QDomNode &node) const
    {
        if (node.isElement() && node.toElement().tagName()=="replace_style_attr")
            return new ThemeSourceReplace(node, this);
        return ThemeDomElement::createChildNode(node);
    }
    ThemeDomNode *ThemePaint::createChildNode(const QDomNode &node) const
    {
        if (node.isElement() && node.toElement().tagName()=="layer")
            return new ThemeLayer(node, this);
        return ThemeDomElement::createChildNode(node);
    }
    ThemeDomNode *Theme::createChildNode(const QDomNode &node) const
    {
        if (node.isElement() && node.toElement().tagName()=="object")
            return new ThemeObject(node, this);
        return ThemeDomElement::createChildNode(node);
    }

    QStringList Theme::objects() {
        QStringList o;
        foreach(ThemeDomNode *c,children()) {
            const ThemeObject *to = dynamic_cast<const ThemeObject*>(c);
            if (to) {
                o << to->getAttr("id");
            }
        }
        return o;
    }


    ThemeModel::ThemeModel()
        : m_modified(false),m_theme(0),m_themeSpecification(0),m_cokoonDocument(0)
    {
        connect(KDirWatch::self(),SIGNAL(dirty(const QString &)),
                this, SLOT(currentFileModifiedOnDisk(const QString &)));
    }

    ThemeModel::~ThemeModel()
    {
        clear();
    }

    void ThemeModel::clear()
    {
        m_currentFileName = QString();
        delete m_theme;
        delete m_themeSpecification;
        delete m_cokoonDocument;
        m_theme = 0;
        m_themeSpecification = 0;
        m_cokoonDocument = 0;
    }

    bool ThemeModel::loadTheme(const QString &fileName)
    {
        clear();

        m_currentFileName = fileName;
        QFile file(fileName);

        if (! file.exists() ) {
            qCritical("File '%s' does not exist.",
                      qPrintable(fileName) );
            return false;
        }

//         if (!file.open(QFile::ReadOnly | QFile::Text)) {
//             qCritical("File '%s' can not be opened: %s", qPrintable(fileName), qPrintable(file.errorString()));
//             return false;
//         }

        if (!m_domDocument.setContent(&file))
            return false;

        m_theme = new Theme(m_domDocument.documentElement(),this);

        QString specResource("cokoon/specs/"+m_theme->spec()+".xml");
        QString specFileName = KStandardDirs::locate("data",specResource);
        qDebug() << specResource << specFileName;

        m_themeSpecification = new DocumentSpecification();
        if (!m_themeSpecification->loadSpecification(specFileName)) {
            return false;
        }

        m_cokoonDocument = new DocumentSpecificationDocument(m_themeSpecification);
//         m_cokoonDocument->loadTheme(fileName);

        activateDirWatch();

        setModified(false);

        // Update TreeItemView
        reset();

        return true;
    }

    bool ThemeModel::save(const QString &fileName)
    {
        deactivateDirWatch();

        m_currentFileName = fileName;
        QFile file(fileName);

        if (!file.open(QFile::WriteOnly | QFile::Text)) {
            qCritical("File '%s' can not be opened: %s", qPrintable(fileName), qPrintable(file.errorString()));
            return false;
        }

        QTextStream stream(&file);
        stream << m_domDocument.toString();

        activateDirWatch();

        return true;
    }

    const QString &ThemeModel::currentFileName() const
    {
        return m_currentFileName;
    }

    Theme *ThemeModel::theme() const
    {
        return m_theme;
    }

    QUndoStack &ThemeModel::undoStack()
    {
        return m_undoStack;
    }

    DocumentSpecification *ThemeModel::specification()
    {
        return m_themeSpecification;
    }

    Document *ThemeModel::cokoonDocument()
    {
        return m_cokoonDocument;
    }

    void ThemeModel::setModified(bool modified)
    {
        m_modified = modified;
        m_cokoonDocument->clear();
        m_cokoonDocument->loadTheme(m_domDocument.toString(), m_currentFileName);
        emit modelModified();
    }

    void ThemeModel::setThemeNodeEdited(const QString &path)
    {
        setModified(true);
        emit themeNodeEdited(path);

        // update TreeItem display..
        QModelIndex idx = getNodeModelIndex(path);
        if(idx.isValid() ) {
            emit dataChanged(idx,idx);
        }
    }

    bool ThemeModel::modified() const
    {
        return m_modified;
    }

    void ThemeModel::currentFileModifiedOnDisk(const QString &path)
    {
        setModified(true);

        if (path == m_currentFileName) {
            int i = KMessageBox::warningYesNo
                (0, i18n("The file %1 was modified by another program.\n\nWhat do you want to do?").arg(path),
                 i18n("File Was Changed on Disk"), KGuiItem(i18n("&Reload File")), KGuiItem(i18n("&Ignore Changes")));

            if ( i == KMessageBox::Yes) {
                loadTheme(path);
            }
        }
    }

    void ThemeModel::activateDirWatch()
    {
        QString f = currentFileName();

        // same file as we are monitoring, return
        if (f == m_dirWatchFile)
            return;

        // remove the old watched file
        deactivateDirWatch();

        // add new file if needed
        if (!f.isEmpty()) {
            KDirWatch::self()->addFile(f);
            m_dirWatchFile = f;
        }
    }

    void ThemeModel::deactivateDirWatch ()
    {
        if (!m_dirWatchFile.isEmpty())
            KDirWatch::self()->removeFile(m_dirWatchFile);

        m_dirWatchFile.clear();
    }

    ThemeDomNode *ThemeModel::getThemeNode(const QString &nodePath) const
    {
        ThemeDomNode *curr = theme();

        const QStringList indices = nodePath.split('/',QString::SkipEmptyParts);
        for ( QStringList::const_iterator it = indices.begin() ; it != indices.end() ; ++it) {
            uint number = (*it).toUInt();
            curr = curr->children().at(number);

            if (!curr)
                break;
        }

        qDebug() << indices << "-> curr:" << curr;

        return curr;
    }
    ThemeDomElement *ThemeModel::getThemeElement(const QString &nodePath) const
    {
        return dynamic_cast<ThemeDomElement*>(getThemeNode(nodePath));
    }


    QModelIndex ThemeModel::getNodeModelIndex(const QString &nodePath) const
    {
        QModelIndex curr = index(0,0,QModelIndex());

        const QStringList indices = nodePath.split('/',QString::SkipEmptyParts);
        for ( QStringList::const_iterator it = indices.begin() ; it != indices.end() ; ++it) {
            uint number = (*it).toUInt();
            curr = index(number,0,curr);

            qDebug() << "for.." << number << "->index:" << curr;

            if (!curr.isValid())
                break;
        }

        qDebug() << indices << "-> curr:" << curr;

        return curr;
    }

//     void ThemeModel::slotModelReset()
//     {
//         reset();
//     }

    QVariant ThemeModel::headerData ( int section, Qt::Orientation orientation, int role ) const
    {
        if (orientation==Qt::Horizontal && role==Qt::DisplayRole) {
            switch(section) {
            case 0:
                return "Element";
            case 1:
                return "ID";
            default:
                return QVariant();
            }
        }
        return QVariant();
    }
    Qt::ItemFlags ThemeModel::flags ( const QModelIndex & index ) const
    {
        if (!index.isValid())
            return Qt::ItemIsEnabled;
        else
            return Qt::ItemIsEnabled|Qt::ItemIsSelectable;
    }
    QVariant ThemeModel::data ( const QModelIndex & index, int role ) const
    {
        if (index.isValid() && role==Qt::DisplayRole) {
            const ThemeDomNode *item = static_cast<const ThemeDomNode *>(index.internalPointer());
            if (item) {
                switch(index.column()) {
                case 0:
                    return item->displayText();
                case 1:
                    return item->displayId();
                default:
                    break;
                }
            }
        }
        return QVariant();
    }
    QModelIndex ThemeModel::index ( int row, int column, const QModelIndex & parent ) const
    {
        if (!m_theme || !theme())
            return QModelIndex();

        if (!parent.isValid()) {
//             theme()->itemModelRow = 0;
//             return createIndex(row,column,theme());
            return theme()->modelIndex(column);
        } else {
            ThemeDomNode *parentItem =
                static_cast<ThemeDomNode *>(parent.internalPointer());
            ThemeDomNode *childItem = parentItem->children().value(row);
            if (childItem) {
//                 childItem->itemModelRow = row;
//                 return createIndex(row,column,childItem);
                return childItem->modelIndex(column);
            }
            else
                return QModelIndex();
        }
    }
    QModelIndex ThemeModel::parent( const QModelIndex & index ) const
    {
        if (index.isValid()) {
            const ThemeDomNode *item = static_cast<const ThemeDomNode*>(index.internalPointer());
            if (item) {
                const ThemeDomNode *parent = item->parent();
                if (parent) {
                    return parent->modelIndex();
//                     eturn createIndex(parent->itemModelRow,0, (void*)parent);
                }
            }
        }
        return QModelIndex();
    }
    int ThemeModel::columnCount ( const QModelIndex & /*parent*/ ) const
    {
        return 2;
    }
    int ThemeModel::rowCount ( const QModelIndex & parent ) const
    {
        if (parent.isValid()) {
            ThemeDomNode *themeItem = static_cast<ThemeDomNode *>(parent.internalPointer());
            if (themeItem)
                return themeItem->children().size();
        }
        return 1;
    }

    bool ThemeModel::deleteNode(const QString &domNodePath)
    {
    }
    bool ThemeModel::moveNodeUp(const QString &domNodePath)
    {
        // TODO
    }
    bool ThemeModel::moveNodeDown(const QString &domNodePath)
    {
    }

//     QDomNode ThemeModel::deleteNodeChild(const QString &parentNodePath, int childIndex)
//     {
//         ThemeDomNode *p = getThemeNode(parentNodePath);
//         if (!p)
//             return false;

//         ThemeDomNode *c = p->children().value(childIndex);
//         if (!c)
//             return false;

//         beginRemoveRows(
//         QDomNode old = p->domNode().removeChild(c->domNode());
//         p->invalidateCache(); c=0; // also deletes c
//         if (old.isNull()) {
//             reset();
//         }

//         return old;
//     }
//     bool ThemeModel::insertNodeChild(const QString &parentNodePath, int childIndex, QDomNode node)
//     {
//     }

}

#include "thememodel.moc"
